File: test_src/cpp/hierarchical_parallelism/atomic/double/target__teams__loop.cpp
#include <iostream>
#include <cmath>
#include <stdexcept>
#




bool almost_equal(double x, double gold, float tol) {
    
        return gold * (1-tol) <= x && x <= gold * ( 1+tol );
    
}

void test_target__teams__loop(){

 // Input and Outputs
 
 const int L = 5;

double counter{};

// Main program

#pragma omp target  map(tofrom:counter) 

{


#pragma omp teams 

{


#pragma omp loop 

    for (int i = 0 ; i < L ; i++ )

{



#pragma omp atomic update

counter += double { 1 };



}

}

}


// Validation
if ( !almost_equal(counter,double { L }, 0.1)  ) {
    std::cerr << "Expected: " << L << " Got: " << counter << std::endl;
    throw std::runtime_error( "target__teams__loop give incorect value when offloaded");
}

}
int main()
{
    test_target__teams__loop();
}

File: test_src/fortran/hierarchical_parallelism/atomic/double_precision/target_teams_loop.f90


program target_teams_loop.f90


    implicit none
  
    INTEGER :: L = 5
    INTEGER :: i
    
    DOUBLE PRECISION :: COUNTER = 0

    
    

     
    
    !$OMP TARGET TEAMS LOOP   MAP(TOFROM: COUNTER) 


    DO i = 1 , L 


    

    

!$OMP ATOMIC UPDATE

counter = counter + 1.


 
     

    END DO

    !$OMP END TARGET TEAMS LOOP
    

    IF  ( ( ABS(COUNTER - L) ) > 10*EPSILON(COUNTER) ) THEN
        write(*,*)  'Expected L Got', COUNTER
        call exit(1)
    ENDIF

end program target_teams_loop.f90